comp/core/agenttelemetry/impl/agenttelemetry_test.go

```diff
@@ -501,6 +505,182 @@ func TestTagAggregateTotalCounter(t *testing.T) {
 	assert.Equal(t, float64(210), m4.Counter.GetValue())
 }
 
+func TestTwoProfilesOnTheSameScheduleGenerateSinglePayload(t *testing.T) {
+	var c = `
+    agent_telemetry:
+      enabled: true
+      profiles:
+        - name: foo
+          metric:
+            metrics:
+              - name: bar.bar
+                aggregate_tags:
+                  - tag1
+        - name: bar
+          metric:
+            metrics:
+              - name: foo.foo
+                aggregate_tags:
+                  - tag1
+    `
+	// setup and initiate atel
+	tel := makeTelMock(t)
+	counter1 := tel.NewCounter("bar", "bar", []string{"tag1", "tag2", "tag3"}, "")
+	counter1.AddWithTags(10, map[string]string{"tag1": "a1", "tag2": "b1", "tag3": "c1"})
+	counter2 := tel.NewCounter("foo", "foo", []string{"tag1", "tag2", "tag3"}, "")
+	counter2.AddWithTags(20, map[string]string{"tag1": "a1", "tag2": "b1", "tag3": "c1"})
+
+	o := convertYamlStrToMap(t, c)
+	s := makeSenderImpl(t, c)
+	r := newRunnerMock()
+	a := getTestAtel(t, tel, o, s, nil, r)
+
+	payloadJSON, err := a.GetAsJSON()
+	assert.NoError(t, err)
+	var payload map[string]interface{}
+	err = json.Unmarshal(payloadJSON, &payload)
+	assert.NoError(t, err)
+
+	// -----------------------
+	// for 2 profiles there are2 metrics, but 1 payload (test is currently payload schema dependent, improve in future)
+
+	// Single payload whcich has sub-payloads for each metric
+	requestType, ok := payload["request_type"]
+	assert.True(t, ok)
+	assert.Equal(t, "agent-metrics", requestType)
+	metricsPayload, ok := payload["payload"].(map[string]interface{})
+	assert.True(t, ok)
+	metrics, ok := metricsPayload["metrics"].(map[string]interface{})
+	assert.True(t, ok)
+	// 2 metrics
+	_, ok = metrics["bar.bar"]
+	assert.True(t, ok)
+	_, ok = metrics["foo.foo"]
+	assert.True(t, ok)
+}
+
+func TestOneProfileWithOneMetricMultipleContextsGenerateTwoPayloads(t *testing.T) {
+	var c = `
+    agent_telemetry:
+      enabled: true
+      profiles:
+        - name: foo
+          metric:
+            metrics:
+              - name: bar.bar
+                aggregate_tags:
+                  - tag1
+    `
+	// setup and initiate atel
+	tel := makeTelMock(t)
+	counter1 := tel.NewCounter("bar", "bar", []string{"tag1", "tag2", "tag3"}, "")
+	counter1.AddWithTags(10, map[string]string{"tag1": "a1", "tag2": "b1", "tag3": "c1"})
+	counter1.AddWithTags(20, map[string]string{"tag1": "a2", "tag2": "b2", "tag3": "c2"})
+
+	o := convertYamlStrToMap(t, c)
+	s := makeSenderImpl(t, c)
+	r := newRunnerMock()
+	a := getTestAtel(t, tel, o, s, nil, r)
+
+	payloadJSON, err := a.GetAsJSON()
+	assert.NoError(t, err)
+	var payload map[string]interface{}
+	err = json.Unmarshal(payloadJSON, &payload)
```

Comment:
Suggestion: What do you think about using `json.Unmarshal` with a [struct](https://gobyexample.com/json) instead of `map[string]interface{}` to avoid lines such as `metricPayloads[0].(map[string]interface{})`?

Reply:
I think I had some problems with that Will try again. Thank you.